src/zenml/integrations/modal/step_operators/modal_step_operator.py

```diff
@@ -163,80 +116,95 @@ def launch(
             environment: The environment variables for the step.
 
         Raises:
-            RuntimeError: If no Docker credentials are found for the container registry.
-            ValueError: If no container registry is found in the stack.
+            RuntimeError: If Modal image construction fails, sandbox creation fails,
+                sandbox execution fails, or Modal application initialization fails.
         """
         settings = cast(ModalStepOperatorSettings, self.get_settings(info))
         image_name = info.get_image(key=MODAL_STEP_OPERATOR_DOCKER_IMAGE_KEY)
         zc = Client()
         stack = zc.active_stack
 
-        if not stack.container_registry:
-            raise ValueError(
-                "No Container registry found in the stack. "
-                "Please add a container registry and ensure "
-                "it is correctly configured."
-            )
-
-        if docker_creds := stack.container_registry.credentials:
-            docker_username, docker_password = docker_creds
-        else:
-            raise RuntimeError(
-                "No Docker credentials found for the container registry."
-            )
-
-        my_secret = modal.secret._Secret.from_dict(
-            {
-                "REGISTRY_USERNAME": docker_username,
-                "REGISTRY_PASSWORD": docker_password,
-            }
+        setup_modal_client(
+            token_id=self.config.token_id,
+            token_secret=self.config.token_secret,
+            workspace=self.config.workspace,
+            environment=settings.modal_environment
+            or self.config.modal_environment,
         )
 
-        spec = modal.image.DockerfileSpec(
-            commands=[f"FROM {image_name}"], context_files={}
-        )
-
-        zenml_image = modal.Image._from_args(
-            dockerfile_function=lambda *_, **__: spec,
-            force_build=False,
-            image_registry_config=modal.image._ImageRegistryConfig(
-                api_pb2.REGISTRY_AUTH_TYPE_STATIC_CREDS, my_secret
-            ),
-        ).env(environment)
+        try:
+            zenml_image = build_modal_image(image_name, stack, environment)
+        except Exception as e:
+            raise RuntimeError(
+                "Failed to build the Modal image from your Docker registry. "
+                "Action required: ensure your ZenML stack's container registry is configured with valid credentials "
+                "and that the base image exists and is accessible. "
```

Comment:
I doubt any of those two cases are actually the cause for failure here? We've built that image just before that pipeline run, and apparently also had permissions to push it there?